shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,317 @@
+    struct When {
+
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+
+            // Tell the director about this when condition
+            reaction->reactor.powerplant.emit(std::make_unique<commands::WhenExpression>(
+                reaction,
+                typeid(State),
+                [](const int& v) { return expr<int>()(v, value); },
+                []() { return NUClear::dsl::operation::CacheGet<State>::get(); },
+                [](NUClear::Reactor& reactor,
+                   std::function<void(const int&)> fn) -> NUClear::threading::ReactionHandle {
+                    return reactor.on<NUClear::dsl::word::Trigger<State>>().then(
+                        [fn](const State& s) { fn(static_cast<int>(s)); });
+                }));
+        }
+    };
+
+    /**
+     * This DSL word is used to create a promise that at some point when running this reaction the state indicated
+     * by T will be true. It may not be on any particular run of the reaction, in which case it will continue to run
+     * that reaction until it is true, or another graph jump becomes necessary.
+     *
+     * @tparam T the condition expression that this reaction will make true before leaving.
+     */
+    template <typename State, enum State::Value value>
+    struct Causing {
+
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+            // Tell the director
+            reaction->reactor.powerplant.emit(
+                std::make_unique<commands::CausingExpression>(reaction, typeid(State), value));
+        }
+    };
+
+    /**
+     * A Task to be executed by the director, and a priority with which to execute this task.
+     *
+     * There are two different types of tasks that can be created using this emit, root level tasks and subtasks.
+     *
+     * Root level tasks:
+     * These are created when a reaction that is not a provider (not a Provides, Entering or Leaving) emits the
+     * task. These tasks form the root of the execution tree and their needs will be met on a highest priority first
+     * basis. These tasks will persist until the provider that they use emits a done task, or the task is re-emitted
+     * with a priority of 0.
+     *
+     * Subtasks:
+     * These are created when a provider task emits a task to complete. These tasks must be emitted each time that
+     * reaction is run to persist as if that reaction runs again and does not emit these tasks they will be
+     * cancelled. Within these tasks the priority is used to break ties between two subtasks that share the same
+     * root task. In these cases the subtask that requested the provider with the highest priority will have its
+     * task executed. The other subtask will be blocked until that task is no longer in the call queue.
```

Comment:
I am confused about the last sentence. Here is my interpretation
```suggestion
     * task executed. The other subtask will be blocked until the first task is no longer in the call queue.
```

Reply:
That would be the correct interpretation. But I think I would change it to be "the active task"

Reply:
perfect